### PR TITLE
Allow clicking on the legend to show/hide plot lines and use more columns

### PR DESCRIPTION
**Pull request by punchagan**
**Before**
image

**After**
image

### CODE REVIEW

**Comment by art-w** [comment]
That's a much better default indeed! What do you think of sorting the labels in desc order? And perhaps collapsing the very small categories together?

<table><tr><td>

```
assemble:0.2680secs
c-linker:2.891secs
cse:0.02300secs
deadcode:0.01100secs
emit:0.02100secs
generate:7.823secs
linearize:0.001000secs
liveness:0.03950secs
parsing:0.6555secs
polling:0.003000secs
regalloc:0.2265secs
scheduling:0.001000secs
selection:0.04100secs
spill:0.03250secs
split:0.01800secs
transl:0.5290secs
typing:4.412secs
```

</td><td>

```
generate:7.823secs
typing:4.412secs
c-linker:2.891secs
(12 grouped):0.6855secs
parsing:0.6555secs
transl:0.5290secs
```

</tr></table>

**Comment by punchagan** [comment]
Great suggestions, thanks!

The sort in descending order of values makes a lot of sense. I think it would also make sense for almost all of the overlaid graphs. There could be some confusion when comparing two versions of the graph (across different branches or some thing like that), when the sort order changes because the values change slightly. But, I don't think this is a common enough case to worry enough about?

The grouping of smaller values as one entry makes sense. I was also wondering if it would make sense to show a total of these also as another "entry", similar to the grouped total of small values.

I'm not sure if the last couple of ideas are generic enough for all benchmarks that we want to add this in the UI, or if we should do this in the `ocaml` benchmark JSON generation step. Do you think overlaid graphs would be used mostly for this kind of a use case where we are splitting up the total value into smaller values and overlaying them on a single graph? May be we should worry about other use cases when we actually hit them?

**Comment by art-w** [comment]
- Yeah, I also think that sorting makes sense for everybody. The order confusion should be rare... and we'll probably work on simplifying the comparison of different branches at some point!
- The grouping is more specific to `ocaml` (it's the only user of overlays right now!), but we can't really do it during the json generation as we don't have the historical values there (what if one metric suddenly degrade and must "exit" the group?) But I think the grouping should be another issue/PR, it's a bit harder to do correctly: do we group on the last values? or do we consider the worst value of each metric over time?

File: frontend/src/LineGraph.res [hunk 1]
```diff
@@ -477,46 +525,72 @@ let make = React.memo((
...
-            ->Belt.Array.get(idx)
-            ->Belt.Option.getWithDefault("#000000")
+            legendColorsByLabel
+            ->Belt.Map.String.getWithDefault(label, "#000000")
             ->Js.String2.substr(~from=1)
+          let active = linesStateByLabel->Belt.Map.String.getWithDefault(label, true)
           <div
             key={idx->Belt.Int.toString}
+            data={label}
+            onClick={handleLegendClick}
```
Comment (art-w) [inline frontend/src/LineGraph.res:573]: Would it make sense to pass `label` to `handleLegendClick` with a partial application? If I understand correctly, the raw javascript `searchDataLabel` is searching the DOM for the above `data={label}`, but then we would'nt need to do that if the handler already had `label`?

File: frontend/src/LineGraph.res [hunk 2]
```diff
@@ -477,46 +525,72 @@ let make = React.memo((
...
+    setLinesState(_ =>
+      linesState->Belt.Array.mapWithIndex((idx, state) => idx == index ? !state : state)
     )
-  let isOverlayed = lastValues->Belt.Array.length > 1
+  }
+  let labels = labels->Belt.Option.getWithDefault([])
+  let lastValues = dataSet->Belt.Array.map(x => x->BeltHelpers.Array.lastExn->DataRow.toValue)
+  let lastValues = Belt.Array.zip(labels, lastValues)->Belt.Map.String.fromArray
+  let linesStateByLabel = Belt.Array.zip(labels, linesState)->Belt.Map.String.fromArray
+  let legendColorsByLabel = Belt.Array.zip(labels, legendColors)->Belt.Map.String.fromArray
```
Comment (art-w) [inline frontend/src/LineGraph.res:544]: I'm not convinced that the `Belt.Map` are really helping here... There are a lot of noisy `Belt.Map...getWithDefault(label, ...)` below to get back the different values associated with a `label`. Would it be possible to structure the data as `(label * (lastValue * lineState * color)) array`? (then all the values would be accessible when we map below)

File: frontend/src/LineGraph.res [hunk 3]
```diff
@@ -336,30 +346,58 @@ let make = React.memo((
...
       data
+      // Dygraph does not display the last tick, so a dummy value
+      // is added a the end of the data to overcome this.
+      // See: https://github.com/danvk/dygraphs/issues/506
       ->Belt.Array.map(x => Belt.Array.concat(x, [DataRow.dummyValue]))
       ->Belt.Array.map(x => x->Belt.Array.map(convertNanToNull))
 
-    let n = (data->Belt.Array.map(Belt.Array.length))[0]
+    let n =
+      data->Belt.Array.map(Belt.Array.length)->Belt.Array.get(0)->Belt.Option.getWithDefault(0)
```
Comment (art-w) [inline frontend/src/LineGraph.res:399]: The `length` could be done as a last step, to skip the `map` :)

Comment (punchagan) [reply frontend/src/LineGraph.res:399]: Ah, right. Thanks!

File: frontend/src/LineGraph.res (hunk 1 above)
Comment (punchagan) [reply frontend/src/LineGraph.res:573]: That's right. Thanks for catching this. :)

File: frontend/src/LineGraph.res (hunk 2 above)
Comment (punchagan) [reply frontend/src/LineGraph.res]: Right, I wasn't very happy with how this code was written too, but hadn't really thought about improving. Thanks for your suggestion.

File: frontend/src/LineGraph.res [hunk 4]
```diff
@@ -322,6 +331,7 @@ let make = React.memo((
   let graphDivRef = React.useRef(Js.Nullable.null)
```
[the remark this answers is not in the document]
Comment (punchagan) [reply frontend/src/LineGraph.res:483]: I've extracted some of the code as separate components.

**Comment by art-w** [comment]
Many thanks, it will really help for the ocaml compiler graphs!

**Pull request merged**